src/cpp/r/session/graphics/RGraphicsDevDesc.cpp

```diff
@@ -1194,15 +1255,136 @@ void releaseMask(SEXP ref, pDevDesc dd)
    switch (engineVersion)
    {
    case 14:
+   case 15:
    default:
-      callback = ((DevDescVersion14*)dd)->releaseMask;
+      callback = ((DevDescVersion15*)dd)->releaseMask;
       break;
    }
    
    if (callback != nullptr)
        callback(ref, dd);
 }
 
+SEXP defineGroup(SEXP source, int op, SEXP destination, pDevDesc dd)
+{
+    int engineVersion = ::R_GE_getVersion();
+
+    SEXP (*callback)(SEXP source, int op, SEXP destination, pDevDesc dd) = nullptr;
+
+    switch (engineVersion)
+    {
+        case 15:
+        default:
+            callback = ((DevDescVersion15*)dd)->defineGroup;
+            break;
+    }
+
+    if (callback != nullptr)
+        return callback(source, op, destination, dd);
+
+    return R_NilValue;
+}
+
+void useGroup(SEXP ref, SEXP trans, pDevDesc dd)
+{
+    int engineVersion = ::R_GE_getVersion();
+
+    void (*callback)(SEXP ref, SEXP trans, pDevDesc dd) = nullptr;
+
+    switch (engineVersion)
+    {
+        case 15:
+        default:
+            callback = ((DevDescVersion15*)dd)->useGroup;
+            break;
+    }
+
+    if (callback != nullptr)
+        callback(ref, trans, dd);
+}
+
+void releaseGroup(SEXP ref, pDevDesc dd)
+{
+    int engineVersion = ::R_GE_getVersion();
+
+    void (*callback)(SEXP ref, pDevDesc dd) = nullptr;
+
+    switch (engineVersion)
+    {
+        case 15:
+        default:
+            callback = ((DevDescVersion15*)dd)->releaseGroup;
+            break;
+    }
+
+    if (callback != nullptr)
+        callback(ref, dd);
+}
+
+void stroke(SEXP path, const pGEcontext gc, pDevDesc dd)
+{
+    int engineVersion = ::R_GE_getVersion();
+
+    void (*callback)(SEXP path, const pGEcontext gc, pDevDesc dd) = nullptr;
+
+    switch (engineVersion)
+    {
+        case 15:
+        default:
+            callback = ((DevDescVersion15*)dd)->stroke;
+            break;
+    }
+
+    if (callback != nullptr)
+        callback(path, gc, dd);
+}
+
+void fill(SEXP path, int rule, const pGEcontext gc, pDevDesc dd)
+{
+    int engineVersion = ::R_GE_getVersion();
+
+    void (*callback)(SEXP path, int rule, const pGEcontext gc, pDevDesc dd) = nullptr;
+
+    switch (engineVersion)
+    {
+        case 15:
+        default:
+            callback = ((DevDescVersion15*)dd)->fill;
+            break;
+    }
+
+    if (callback != nullptr)
+        callback(path, rule, gc, dd);
+}
+
+void fillStroke(SEXP path, int rule, const pGEcontext gc, pDevDesc dd)
+{
+    int engineVersion = ::R_GE_getVersion();
+
+    void (*callback)(SEXP path, int rule, const pGEcontext gc, pDevDesc dd) = nullptr;
+
+    switch (engineVersion)
+    {
+        case 15:
+        default:
+            callback = ((DevDescVersion15*)dd)->fillStroke;
+            break;
+    }
+
+    if (callback != nullptr)
+        callback(path, rule, gc, dd);
+}
+
+SEXP capabilities(SEXP cap)
+{
+    SEXP (*callback)(SEXP cap) = nullptr;
+
+    if (callback != nullptr)
+    return callback(cap);
```

Comment:
nit: indention here?

Reply:
It also looks like we're not reading the value of the `callback` pointer similar to what we're doing in other methods; does that need to be added back here as well?

Reply:
Yeah good call, I was not exactly sure what to do here because in 
[the R repo code](https://github.com/wch/r-source/blob/68f448578b206175d5d10eb4bd1c55b4ca4a0227/src/include/R_ext/GraphicsDevice.h#L787), the `capabilities` signature doesn't take `pDevDesc dd` like all the other methods do, so wasn't sure what to do in place of 
```
switch (engineVersion)
    {
        case 15:
        default:
            callback = ((DevDescVersion15*)dd)->capabilities;
            break;
    }
```
given that `dd` is not being passed in

Reply:
Good question... I think we can visit this in a separate PR; having this be a no-op for now is probably fine.